overlord/state.go

```diff
@@ -20,25 +20,49 @@
 package overlord
 
 import (
+	"encoding/json"
+	"fmt"
 	"io"
 )
 
 // State represents a snapshot of the system state.
-type State struct{}
+type State struct {
+	entries map[string]json.RawMessage
+}
+
+// NewState returns an empty system state.
+func NewState() *State {
+	return &State{
+		entries: make(map[string]json.RawMessage),
+	}
+}
 
 // Get unmarshals the stored value associated with the provided key
 // into the value parameter.
 func (s *State) Get(key string, value interface{}) {
+	err := json.Unmarshal(s.entries[key], value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not retrieve and unmarshal state entry %q: %v", key, err))
```

Comment:
I wonder if this should even be a panic. It seems like the outcome of something assuming facts about the state which change with future versions may be hard to get precisely right every time. Might be better for us to be less confident about our own ability to get it right every time, and instead have a slightly softer failure mode for when it happens.